Add FuzzySet Class for Triangular Fuzzy Sets

Shreya123714 opened the pull request:
**###hacktoberfest**

### Description:
This pull request introduces the FuzzySet class, designed to represent and manipulate triangular fuzzy sets widely used in fuzzy logic and control systems. The creation of this class was motivated by the need for a versatile tool for fuzzy logic operations.

### Features:
- **FuzzySet Class**: The FuzzySet class is defined with attributes such as name, a (left boundary), b (peak), and c (right boundary) to describe the fuzzy set. It also provides methods for membership calculation, union, intersection, and complement operations.

### Functionality:
- **Membership Calculation**: The membership method calculates the membership value of an input x in the fuzzy set, handling cases where x is outside the fuzzy set boundaries.

- **Union and Intersection**: The union and intersection methods allow users to calculate the union and intersection of fuzzy sets, respectively. These operations create new fuzzy sets based on the input fuzzy sets.

- **Complement (Negation)**: The complement method computes the complement (negation) of the fuzzy set, making it a valuable addition for fuzzy logic operations.

- **Plotting**: The plot method generates a visual representation of the membership function for effective visualization and analysis.

### Testing:
I have added a comprehensive test suite for the FuzzySet class to ensure its correctness and robustness. The tests cover various scenarios, including both valid and erroneous input values.

### Example Usage:
The example usage section demonstrates how to create FuzzySet instances, perform operations, and visualize the membership functions of fuzzy sets. It provides practical insights into the class's capabilities.

I welcome your review and feedback, and I'm open to further improvements. Transparency is important, and I have designed this class based on existing knowledge and references from reputable sources.

Shreya123714 commented:
wait I will fix this..

Shreya123714 commented on fuzzy_logic/fuzzy_operations.py:
```diff
@@ -0,0 +1,160 @@
+
+        plt.plot(x, y, label=self.name)
+
+    def __str__(self):
```
Ok

Shreya123714 commented:
Hey can you put a Hacktoberfest label into this pull request??

cclauss commented:
Please give me time to review the code without changing it.

Shreya123714 commented:
> Please give me time to review the code without changing it.

Hey, while you do that , can I fix few typo error , and make the code more descriptive, 
I am new to open source , now after reading the bot's suggestion i found few loop holes like " function hint "problem,

cclauss reviewed (approved):
Looks cool but I do not get all the details...

I moved some tests from fuzzy_logic/test_fuzzy_logic.py into doctests and removed that file.  I did several updates to the main file to make it pass our tests.  If it does so then I will approve and merge it.  If you see missing tests or other errors, please open a new pull request with changes and @mention me on it.

Shreya123714 commented:
> Looks cool but I do not get all the details...
> 
> I moved some tests from fuzzy_logic/test_fuzzy_logic.py into doctests and removed that file. I did several updates to the main file to make it pass our tests. If it does so then I will approve and merge it. If you see missing tests or other errors, please open a new pull request with changes and @mention me on it.


same , my PR started failing, once I added that test_fuzzy_logic.py , i did so to remove the " required test " label, 
But sure , I shall work on it 😄😄😄(●'◡'●)  Thank you for merging the PR , Thank you Thank you Thank You so MUCH!!!!!!!!!!!!

cclauss commented:
pytest not finding tests is often about not having an empty `__init__.py` file in the same directory.  In this repo, we tend to use doctest because the tests live with the code and it is often easier for new contributions to understand doctests.

Please feel free to create pull requests to make this code what you want it to be.

cclauss commented:
Also, Python `set()`
* https://docs.python.org/3/library/stdtypes.html#set
* https://docs.python.org/3/tutorial/datastructures.html#sets

Enable the following notation: 
```
x in set  # **membership** but must return a bool --> __contains__(self, item)
set & other  # **intersection** --> __and__(self, other)
set | other  # **union** --> __or__(self, other)
~set  # **complement** --> __invert__(self)

If you implemented `FuzzySet.__contains__()` you could enable `x in fuzzy_set` for membership.
If you implemented `FuzzySet.__and__()` you could enable `fuzzy_set_a & fuzzy_set_b` for intersection.
If you implemented `FuzzySet.__and__()` you could enable `fuzzy_set_a | fuzzy_set_b` for union.
If you implemented `FuzzySet.__contains__()` you could enable `~fuzzy_set` for complement.
